cpp/csp/python/PyEngine.h

```diff
@@ -54,6 +55,24 @@ class CSPIMPL_EXPORT PyEngine final: public PyObject
     Engine    * m_engine;
 };
 
+// Generic engine shutdown function for push-type adapters
+template<typename T>
+PyObject * PyEngine_shutdown( T * self, PyObject * args, PyObject * kwargs )
+{
+    CSP_BEGIN_METHOD;
+
+    PyObject * pyException;
+    if( !PyArg_ParseTuple( args, "O", &pyException ) )
+        return NULL;
+
+    if( !PyExceptionInstance_Check( pyException ) )
+       CSP_THROW( TypeError, "Expected Exception object as argument for shutdown_engine: got " << Py_TYPE( pyException ) -> tp_name );
```

Comment:
Why are we throwing here, we need to call `shutdown`?

Reply:
Good catch, I changed it in PyAdapterManager, forgot here. Fixed